### ISSUE TITLE
Improve platform command line length limits

### ISSUE THREAD

**Issue by antonbabenko** [pre-commit/pre-commit#691]
I have a question related to #510 

When I run `pre-commit run` all hooks work as expected (processing just changed files), but `pre-commit run -a` divides a list of files into 5 bulks and run the same hook in parallel executions. This creates unnecessary checks of files which belong to the same directory.

Is it possible to control that certain hooks should not be executed in parallel but rather pass all filenames to it?

Alternatively, is there an environment variable or something like that which says that pre-commit is running a hook for all files?

**Comment by asottile** [comment]
So pre-commit doesn't _yet_ run anything in parallel (as proposed in #510 -- and there will be a way to opt out of it once implemented).

What you're probably seeing is the batching of calls.  Passing all arguments to commands will not work in in all situations (there is a command length limit!).  On unixlike systems, coreutils provides a utility [`xargs`](http://man7.org/linux/man-pages/man1/xargs.1.html) to compute how many arguments to pass to avoid this limit.   pre-commit implements a similar helper but is [extremely timid](https://github.com/pre-commit/pre-commit/blob/51ac0e8bc0a1b61939283254c4bd5189c919f919/pre_commit/xargs.py#L8-L10) due to limits found on windows.

This limit can probably be increased on platforms where the limit is higher (if this can be detected?) and then you'll be much more likely to get one in a single batch.  I'd love a PR that does this :D

**Comment by chriskuehl** [comment]
@antonbabenko no hooks operate in parallel yet; that issue is requesting the functionality, but it hasn't been implemented yet.

It's true that pre-commit may split files across multiple hook invocations if there are too many files, but this happens in serial and is sort of unavoidable (since there are limits on the length of `argv`, see `xargs --show-limits`; though pre-commit is a lot more conservative than what the actual max probably is on your system):

https://github.com/pre-commit/pre-commit/blob/51ac0e8bc0a1b61939283254c4bd5189c919f919/pre_commit/xargs.py#L8-L10

**Comment by antonbabenko** [comment]
Thanks for the explanation. I am fine with the little overhead when running the same hook over the same directory.

**Comment by asottile** [comment]
The code hints here are:

- [`os.sysconf`](https://docs.python.org/3/library/os.html#os.sysconf)
- [`SC_MAX_ARG`](ftp://ftp.gnu.org/old-gnu/Manuals/glibc-2.2.3/html_chapter/libc_31.html#SEC648)
- [`xargs` implementation](http://git.savannah.gnu.org/cgit/findutils.git/tree/xargs/xargs.c)
    - `xargs` leaves room for posix mandated `2048` bytes, environ, and command.  It maxes by default at 128KiB (`131072` bytes).

And that none of this can be trusted on windows heh (so probably only do this if `sys.platform != 'win32'` (should also double check how `cygwin` responds here).

**Comment by ikonst** [comment]
To clarify above, "by default" means "unless you specify `-s max-chars`".

**Comment by georgeyk** [comment]
hey @asottile,

cygwin seems to set 32kb (https://cygwin.com/git/gitweb.cgi?p=newlib-cygwin.git;a=blob;f=winsup/cygwin/include/limits.h;h=3550c4fcb82589062b410b4f9b0c4e5fa09eb2e8;hb=28ecec475ff423b368bcca329f42cfed29308d61#l173)

So, should pre-commit/xargs implementation consider a value between 32k and 128k (+ 4kb minimum) ?

Another refactor in pre-commit/xargs.py is to stop using str length and use the "lenght" in bytes.

**Comment by asottile** [comment]
makes sense to me, I'd try and follow `xargs` as close as possible!  using `bytes` probably makes sense as well (the limit on windows is actually characters because the api takes wide chars directly, yay)

**Comment by georgeyk** [comment]
Here's summary (just for future reference) https://blogs.msdn.microsoft.com/oldnewthing/20031210-00/?p=41553/

I was looking `xargs` code and they do a lot of things to compute the correct value. Maybe we could take a shortcut ?
Something like:
```
import os
import sys


def _get_max_arg_length():
    headroom = 2 * 1024
    min_arg_length = 4 * 1024
    initial_max_arg_length = 128 * 1024

    if sys.platform == 'win32':
        initial_max_arg_length = 32 * 1024

    max_arg = initial - sys.getsizeof(os.environ) - headroom
    # xxx: sys.getsizeof just to illustrate the idea

    return max(min_arg_length, max_arg)

```

**Comment by asottile** [comment]
makes sense to me!   probably use `os.sysconf` on posix to get the `128 * 1024` value but other than that seems about right

**Comment by georgeyk** [comment]
Just one more thing. the output of `xargs --show-limits`:
```
POSIX upper limit on argument length (this system): 2088635
POSIX smallest allowable upper limit on argument length (all systems): 4096
Maximum length of command we could actually use: 2082166
Size of command buffer we are actually using: 131072
```

So, `SC_ARG_MAX` is 2088635 (200k) and we should use the maximum value possible or consider 128k as "safe-upper-limit" ?

**Comment by asottile** [comment]
I forget how xargs makes its decision but I think the smaller value makes sense (128k)

**Comment by ushuz** [comment]
We're running into `ArgumentTooLongError` on Travis CI when invoking pre-commit through make.

Our makefile looks like this:

```makefile
lint:
  ci/lint.sh
```

Run `ci/lint.sh` directly won't have any trouble, but `make lint` will raise `ArgumentTooLongError`.

**Comment by asottile** [comment]
Do you have a repo I can peek at? For now use 1.14.4, I'll get a fix in for 1.15.x

**Comment by ushuz** [comment]
@asottile Sorry, private repo. Here's a screenshot of a debug build on Travis CI:

![image](https://user-images.githubusercontent.com/1430856/55335109-16428d80-54cd-11e9-8bfc-75030d45a011.png)

The Makefile I used:

```makefile
lint2:
  pre-commit --files Makefile
```

Dug deeper into code, it seems that `os.sysconf(str('SC_ARG_MAX'))` returned `-1`.

**Comment by asottile** [comment]
@ushuz that's perfect! And this is Linux I assume?

**Comment by asottile** [comment]
@ushuz I've released 1.15.1 which should have a fix for this
